### PR TITLE
Feature/mergeplatformdirs

### PR DESCRIPTION
**Pull request by KlausPopp**
- [x] Add moducop cpu01 support
- [x] Use only single "yocto" directory for all types of builds 
- [x] Add "bringup" image types
- [x] Replace tdx-reference-minimal-image by edgefarm images
- [x] Move to tdx bsp 5.1
- [x] Add networkmanager, modemmanager to bringup+edgefarm images
- [x] Changed versioning for local builds (if dirty, use developer name+date)
- [x] ci pipeline: auto push to mender

### CODE REVIEW

File: yocto/ci/config-dev.yaml [hunk 1]
```diff
@@ -2,24 +2,25 @@
...
+# Set this via fly .... -v name=cpu01-bringup (or to your desired target)
+#name: cpu01-bringup
 
 # Git configuration
 # -----------------
 ## Specify the git repository to work with
 git_source: git@github.com:ci4rail/ci.os.lmp.git
 ## Specify the git branch to work with
-git_branch: main
+git_branch: feature/mergeplatformdirs
```
Comment (siredmar) [inline yocto/ci/config-dev.yaml:13]: I guess this should be main or something releasable

File: yocto/ci/config-dev.yaml [hunk 2]
```diff
@@ -2,24 +2,25 @@
 # Project configuration
 # -----------------------------
 ## Defines the name for the project
-name: mender-on-verdin
+# Set this via fly .... -v name=cpu01-bringup (or to your desired target)
+#name: cpu01-bringup
```
Comment (siredmar) [inline yocto/ci/config-dev.yaml:6]: Give an useful default name.

File: yocto/dobi.yaml [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+    - mount-buildingblock-gitversion-source-dir
+  sources: yocto/layer-revs
+  artifact: /git/yocto/.generated-cpu01-edgefarm-image-naming.yaml
+  user: "{user.uid}:{user.gid}"
+  env:
+    - "USER={user.name}"
+
+
+# CPU01-edgefarm: Interactive yocto shell
+# Warning: Image/Mender artifact names will be defaults!
```
Comment (siredmar) [inline yocto/dobi.yaml:155]: To all those warnings: 
They will be defaults unless you override them with the mender.env file.
So please rephrase this a little to make that clear or delete the comment.

File: yocto/dobi.yaml [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+  env:
+    - USER_ID={user.uid}
+    - GROUP_ID={user.gid}
+    - TERM=xterm-256color
+    - SHELL=/bin/bash
+    - MENDER_SERVER_URL={env.MENDER_SERVER_URL}
+    - MENDER_TENANT_TOKEN={env.MENDER_TENANT_TOKEN}
+    - IMAGE_NAME_SUFFIX={env.IMAGE_NAME_SUFFIX}
+  annotations:
+    description: "Build verdindev-edgefarm-image"
```
Comment (siredmar) [inline yocto/dobi.yaml:289]: Indent descriptions like this
```
Resources:
  build-all                      --> [alias] build all images
  cpu01-bringup-build-image      --> Build cpu01-bringup-image
  cpu01-bringup-yocto-shell      --> interactive yocto shell cpu01-bringup config
  cpu01-edgefarm-build-image     --> Build cpu01-edgefarm-image
  cpu01-edgefarm-yocto-shell     --> interactive yocto shell cpu01-edgefarm config
  verdindev-edgefarm-build-image --> Build verdindev-edgefarm-image
  verdindev-edgefarm-yocto-shell --> interactive yocto shell verdindev-edgefarm config
  yocto-clean                    --> Cleanup
```

**Review by siredmar — CHANGES_REQUESTED** [review]
First quick review.

File: yocto/layer-revs [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+meta-qt5: 0d8eb956015acdea7e77cd6672d08dce18061510
+meta-rust: e4d25b98083bcecb94df6ee189a165d63ede7f3d
+meta-toradex-bsp-common: cdbfe017be9096adbbfb4d3bbea0421e44f54cf3
+meta-toradex-demos: 38705bdc1ff671597680c5a9c9d1a76822bb5620
+meta-toradex-distro: 48913ab629a231913e3e147e93e57f7cf6019ac6
+meta-toradex-nxp: 8e762e3c8ec5c84f76254f442f37c70dfb7e1b68
+meta-toradex-tegra: b214067b27204f471cf405a1e3997d3da233a8f9
+meta-virtualization: ff997b6b3ba800978546098ab3cdaa113b6695e1
+meta-yocto: c1c79c1a122b053fb3ce0187b3cc89df1bb3bdd6
+openembedded-core: b885888df67eb5cdb3b82f4f0a07369a449e223b
```
[the remark this answers is not in the document]
Comment (KlausPopp) [reply yocto/layer-revs]: Yes, it is generated. But it must be part of the repo, as it records which layers were used for building.

File: yocto/TODO [hunk 6]
```diff
@@ -0,0 +1,11 @@
+Remove
```
[the remark this answers is not in the document]
Comment (KlausPopp) [reply yocto/TODO:1]: OK

File: yocto/ci/config-dev.yaml [hunk 7]
```diff
@@ -2,24 +2,25 @@
...
 ## This enables/disables ssl verification of the git resource
 git_skip_ssl_verification: false
 
 # S3 configuration
 # ----------------
 ## Specify the S3 endpoint URI
-s3_endpoint: http://172.17.0.1:9000
+s3_endpoint: https://minio.ci4rail.com
 ## This enables/disables the ssl verification for the S3 endpoint
 s3_skip_ssl_verification: false
```
[the remark this answers is not in the document]
Comment (KlausPopp) [reply yocto/ci/config-dev.yaml:26]: There is just one pipeline file. The different pipelines are generated via multiple calls to fly, using the -v name=<machine>-<image> parameter, as documented in the README

File: yocto/kasfile-iotedge.yaml [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  src/meta-rust:
+  # fixed revision e4d25b98 from branch `master`
+    url: git://github.com/meta-rust/meta-rust.git
+    refspec: e4d25b98083bcecb94df6ee189a165d63ede7f3d
+    
+  # ci4rail
+  src/meta-ci.os:
+    url: https://github.com/ci4rail/meta-ci.os.git
+    refspec: feature/moducop
+
```
[the remark this answers is not in the document]
Comment (KlausPopp) [reply yocto/kasfile-iotedge.yaml:24]: Ups. You are right. Deleted

File: yocto/layer-revs (hunk 5 above)
[the remark this answers is not in the document]
Comment (siredmar) [reply yocto/layer-revs]: What if i use different refspecs for another marchine?

File: yocto/dobi.yaml [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+  env:
+    - USER_ID={user.uid}
+    - GROUP_ID={user.gid}
+    - TERM=xterm-256color
+    - SHELL=/bin/bash
+    - MENDER_SERVER_URL={env.MENDER_SERVER_URL}
+    - MENDER_TENANT_TOKEN={env.MENDER_TENANT_TOKEN}
+    - IMAGE_NAME_SUFFIX={env.IMAGE_NAME_SUFFIX}
+  annotations:
+    description: "interactive yocto shell cpu01-bringup config"
```
[the remark this answers is not in the document]
Comment (KlausPopp) [reply yocto/dobi.yaml]: Done

File: yocto/dobi.yaml [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+job=yocto-layer-revparse:
+  use: image-kas
+  entrypoint: /git/layer-revparse.sh /git/yocto/src /git/yocto/layer-revs
+  mounts:
+    - mount-buildingblock-gitversion-source-dir
+  sources: yocto/src
+  artifact: /git/yocto/layer-revs
+  user: "{user.uid}:{user.gid}"
+  #annotations:
+  #  description: "get layer revisions"
```
[the remark this answers is not in the document]
Comment (KlausPopp) [reply yocto/dobi.yaml]: Removed.

File: yocto/dobi.yaml [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+    /git/yocto/.generated-cpu01-bringup-image-naming.yaml
+  mounts:
+    - mount-buildingblock-gitversion-source-dir
+  sources: yocto/layer-revs
+  artifact: /git/yocto/.generated-cpu01-bringup-image-naming.yaml
+  user: "{user.uid}:{user.gid}"
+  env:
+    - "USER={user.name}"
+  #annotations:
+  #  description: "Generate cpu01-bringup image version"
```
[the remark this answers is not in the document]
Comment (KlausPopp) [reply yocto/dobi.yaml]: Removed.

File: yocto/dobi.yaml [hunk 12]
```diff
@@ -0,0 +1,310 @@
+# ===================================================
```
[the remark this answers is not in the document]
Comment (KlausPopp) [reply yocto/dobi.yaml:4]: Removed

File: yocto/dobi.yaml [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+    /git/yocto/layer-revs
+    cpu01-edgefarm
+    /git/yocto/.generated-cpu01-edgefarm-image-naming.yaml
+  mounts:
+    - mount-buildingblock-gitversion-source-dir
+  sources: yocto/layer-revs
+  artifact: /git/yocto/.generated-cpu01-edgefarm-image-naming.yaml
+  user: "{user.uid}:{user.gid}"
+  env:
+    - "USER={user.name}"
```
[the remark this answers is not in the document]
Comment (KlausPopp) [reply yocto/dobi.yaml]: No. The variable substitution in the entrypoint string did not work

File: yocto/dobi.yaml [hunk 14]
```diff
@@ -0,0 +1,310 @@
...
+  env:
+    - USER_ID={user.uid}
+    - GROUP_ID={user.gid}
+    - TERM=xterm-256color
+    - SHELL=/bin/bash
+    - MENDER_SERVER_URL={env.MENDER_SERVER_URL}
+    - MENDER_TENANT_TOKEN={env.MENDER_TENANT_TOKEN}
+    - IMAGE_NAME_SUFFIX={env.IMAGE_NAME_SUFFIX}
+  annotations:
+    description: "interactive yocto shell verdindev-edgefarm config"
```
[the remark this answers is not in the document]
Comment (KlausPopp) [reply yocto/dobi.yaml]: Really? We have only 8 jobs in total. Its not necessary right now

File: yocto/kasfile-cpu01-bringup.yaml [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+distro: tdx-xwayland
+# The target that is built by bitbake
+target: ci4rail-bringup-image
+
+repos:
+    
+  # ci4rail
+  src/meta-ci.os:
+    url: https://github.com/ci4rail/meta-ci.os.git
+    refspec: feature/moducop
```
[the remark this answers is not in the document]
Comment (KlausPopp) [reply yocto/kasfile-cpu01-bringup.yaml:28]: Yes. Lets do this as the last step before we merge to main

File: yocto/kasfile-cpu01-edgefarm.yaml [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+distro: tdx-xwayland
+# The target that is built by bitbake
+target: ci4rail-edgefarm-image
+
+repos:
+    
+  # ci4rail
+  src/meta-ci.os:
+    url: https://github.com/ci4rail/meta-ci.os.git
+    refspec: feature/moducop
```
[the remark this answers is not in the document]
Comment (KlausPopp) [reply yocto/kasfile-cpu01-edgefarm.yaml:29]: Yes. Lets do this as the last step before we merge to main

File: gen-image-version.sh [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+#   VERSION=dirty_<fullsemver>.<branch><user>.<datetime>
+# Else
+#   VERSION=<fullsemver>.<branch>.<shortsha>
+#
+# To be called with arguments: <git-root> <layer-refs file> <outfile>
+#
+# This script is expected to be called by dobi
+
+set -e
+
```
[the remark this answers is not in the document]
Comment (KlausPopp) [reply gen-image-version.sh:18]: done

File: gen-image-version.sh [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+    exit 1
+fi
+
+git_root=${1}
+layer_refs=${2}
+mender_prefix=${3}
+out_file=${4}
+
+is_dirty=0
+
```
[the remark this answers is not in the document]
Comment (KlausPopp) [reply gen-image-version.sh:31]: Done

File: yocto/ci/config-dev.yaml (hunk 1 above)
Comment (KlausPopp) [reply yocto/ci/config-dev.yaml:13]: Yes. Lets do this as the last step before we merge to main

File: yocto/ci/config-dev.yaml (hunk 2 above)
Comment (KlausPopp) [reply yocto/ci/config-dev.yaml:6]: No. it Should fail if no name is set via fly -v.

File: yocto/dobi.yaml (hunk 3 above)
Comment (KlausPopp) [reply yocto/dobi.yaml]: Rephrased

File: yocto/dobi.yaml (hunk 4 above)
Comment (KlausPopp) [reply yocto/dobi.yaml]: Done

File: yocto/kasfile-cpu01-bringup.yaml [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+
```
[the remark this answers is not in the document]
Comment (KlausPopp) [reply yocto/kasfile-cpu01-bringup.yaml:29]: Ok, moved to meta-ci.bringup

File: yocto/ci/config-dev.yaml (hunk 2 above)
Comment (siredmar) [reply yocto/ci/config-dev.yaml:6]: I see. I don't like it :) 
There is a section in the readme that explains it a little bit, but this is far from: "hey, i'm new here and i know exactly what i am doing". I know that setting pipelines can be a one shot thing, but then at least provide a shell script that runs fly in a loop for all the different pipelines. If no changes in the pipelines are present, then fly would tell you so.

File: yocto/ci/config-dev.yaml (hunk 1 above)
Comment (siredmar) [reply yocto/ci/config-dev.yaml:13]: Ok

File: yocto/ci/config-dev.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (siredmar) [reply yocto/ci/config-dev.yaml:26]: See my comment above.

File: yocto/dobi.yaml [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+    - yocto/install/images/moducop-cpu01
+  annotations:
+    description: "\t--> Build verdindev-edgefarm-image"
+
+
+# Clean 
+job=yocto-clean:
+  use: image-kas
+  entrypoint: /bin/bash -c "rm -rf /downloads/* /sstate-cache/* /install/* /work/build/*"
+  #entrypoint: /bin/bash
```
Comment (siredmar) [inline yocto/dobi.yaml:226]: remove this line

File: yocto/kasfile-cpu01-bringup.yaml (hunk 15 above)
[the remark this answers is not in the document]
Comment (siredmar) [reply yocto/kasfile-cpu01-bringup.yaml:28]: ok

File: yocto/kasfile-cpu01-bringup.yaml (hunk 19 above)
[the remark this answers is not in the document]
Comment (siredmar) [reply yocto/kasfile-cpu01-bringup.yaml:29]: nice!

File: yocto/kasfile-cpu01-edgefarm.yaml (hunk 16 above)
[the remark this answers is not in the document]
Comment (siredmar) [reply yocto/kasfile-cpu01-edgefarm.yaml:29]: ok

File: yocto/pipeline.yaml [hunk 21]
```diff
@@ -174,17 +181,32 @@ jobs:
...
+          inputs:
+            - name: install
+          run:
+            path: /bin/sh
+            args:
+              - -c
+              - |
+                /mender-cli login --username ((HOSTED_MENDER_USER_EMAIL)) --password ((HOSTED_MENDER_PASSWORD));
+                ls -l install;
+                /mender-cli artifacts upload install/*.mender;
```
Comment (siredmar) [inline yocto/pipeline.yaml:202]: really nice! :+1:

File: yocto/layer-revs (hunk 5 above)
[the remark this answers is not in the document]
Comment (KlausPopp) [reply yocto/layer-revs]: good catch, but I think its not that problematic. Kas will checkout the refspecs you specify in the kasfile. 
So, under src, you will always have the refspecs of the last build.
The only problem would be if you attempt to build multiple images with conflicting refspecs in parallel.
The content of the layer-revs file is not important as the file is not checked in anymore. It is only used to check if any layer is dirty. 
From my point of view, it's ok like this

File: yocto/ci/config-dev.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (KlausPopp) [reply yocto/ci/config-dev.yaml:26]: Which comment above?

[the remark this answers is not in the document]
Comment (siredmar) [reply yocto/ci/config-dev.yaml:26]: regarding the other machines using `-v name=...`

[the remark this answers is not in the document]
Comment (KlausPopp) [reply yocto/ci/config-dev.yaml:26]: Added set-pipelines.sh

File: yocto/ci/config-dev.yaml (hunk 2 above)
Comment (KlausPopp) [reply yocto/ci/config-dev.yaml:6]: Added set-dev-pipelines.sh

File: yocto/dobi.yaml (hunk 20 above)
Comment (KlausPopp) [reply yocto/dobi.yaml:226]: Done

**Review by siredmar — APPROVED** [review]

**Pull request merged**